doc/man/ncftpls.1
.TH ncftpls 1 "NcFTP Software" ncftpls
.SH NAME
ncftpls - Internet file transfer program for scripts
.SH "SYNOPSIS"
.PP
.B ncftpls
.RI [ "options" ]
.I "ftp://url.style/host/path/name/"
.\"-------
.SH "OPTIONS"
.\"-------
.SS
Command line flags:
.TP 8
.B \-1
Most basic format, one item per line.
.TP 8
.B \-l
Long list format.
.TP 8
.B \-R
Long list format, recurse subdirectories.
Equivalent to "-x -lR".
.TP 8
.BI "-x -" "XX"
Additional
.I ls
flags to pass on to the server.
.TP 8
.BI "-u " "XX"
Use username
.I XX
instead of anonymous.
.TP 8
.BI "-p " "XX"
Use password
.I XX
with the username.
.TP 8
.BI "-P " "XX"
Use port number
.I XX
instead of the default FTP service port (21).
.TP 8
.BI "-d " "XX"
Use the file
.I XX
for debug logging.
.TP 8
.BI "-t " "XX"
Timeout after
.I XX
seconds.
.TP 8
.B -E
Use regular (PORT) data connections.
.TP 8
.B -F
Use passive (PASV) data connections.
The default is to use passive, but to fallback to
regular if the passive connection fails or times out.
.TP 8
.BI "-r " "XX"
Redial a maximum of 
.I XX
times until connected to the remote FTP server.
.TP 8
.BI "-W " "XX"
Send raw FTP command
.I XX
after logging in.
.TP 8
.BI "-X " "XX"
Send raw FTP command
.I XX
after each file transferred.
.TP 8
.BI "-Y " "XX"
Send raw FTP command
.I XX
before logging out.
.IP
The
.BR "-W" ", " "-X" ", and " "-Y"
options are useful for advanced users who need to tweak
behavior on some servers.
For example, users accessing mainframes might need to send
some special SITE commands to set blocksize and record format information.
.IP
For these options, you can use them multiple times each if you need
to send multiple commands.
For the
.B "-X"
option, you can use the cookie
.B %s
to expand into the name of the file that was transferred.
.\"-------
.SH "DESCRIPTION"
.\"-------
.PP
The
purpose of
.I ncftpls
is to do remote directory listings using
the File Transfer Protocol without entering an interactive shell.
This lets you write shell scripts or other unattended
processes that can do FTP.
.PP
The default behavior is to print the directory listing in columnized
format (i.e. ls\ \-CF), but that is not very useful for scripting.
This example uses the
.B \-1
flag, to print one file per line:
.RS 4
.br
.sp
$ ncftpls -1 ftp://ftp.ncftp.com/pub/ncftp/
.RE
.PP
You can also do a remote "ls\ \-l", by using "ncftpls\ \-l".
If you want to try other flags, you have to use them with the
.B \-x
flag.
For example, if you wanted to do a remote "ls\ \-lrt", you could
do this:
.RS 4
.br
.sp
$ ncftpls -x "-lrt" ftp://ftp.ncftp.com/pub/ncftp/
.RE
.PP
By default the program tries to open the remote host
and login anonymously, but you can specify a username
and password information like you can with
.I ncftpget
or
.IR ncftpput "."
.\"-------
.SH "DIAGNOSTICS"
.\"-------
.PP
.I ncftpls
returns the following exit values:
.TP 8
0
Success.
.TP 8
1
Could not connect to remote host.
.TP 8
2
Could not connect to remote host - timed out.
.TP 8
3
Transfer failed.
.TP 8
4
Transfer failed - timed out.
.TP 8
5
Directory change failed.
.TP 8
6
Directory change failed - timed out.
.TP 8
7
Malformed URL.
.TP 8
8
Usage error.
.TP 8
9
Error in login configuration file.
.TP 8
10
Library initialization failed.
.TP 8
11
Session initialization failed.
.\"-------
.SH "AUTHOR"
.\"-------
.PP
Mike Gleason, NcFTP Software (http://www.ncftp.com).
.\"-------
.SH "SEE ALSO"
.\"-------
.PP
.IR ncftpput (1),
.IR ncftpget(1),
.IR ncftp (1),
.IR ftp (1),
.IR rcp (1),
.IR tftp (1).
.PP
.IR "LibNcFTP" " (http://www.ncftp.com/libncftp/)."


CHANGELOG
Change Log:
===========

3.0.1, 2000-03-27:

+ ncftpput has been changed to cd to the destination directory on the
  remote host before storing files.  In addition, it does it one
  subdirectory node at a time, like URL mode does for ncftpget.

+ Bug fixed where the "ncftp_bookmarks --dimensions-terse" hack could
  be closed and generate a SIGPIPE which caused it to exit uncleanly,
  and possibly leave the terminal state hosed.

+ Added a snippet from Felix von Leitner <leitner@fefe.de> to get the
  terminal width fairly painlessly, so the aforementioned hack should
  not be necessary in many cases.

+ Dtterm and vt220 added to the short list of vt100ish term types.

+ Win32 version bug fixed where ncftpbookmarks was not honoring the
  HOME enviornment variable, instead writing data to the install
  directory under Program Files.

+ Win32 version now uses Wise installer instead of InstallShield.


3.0.0, 2000-03-19:

+ New "redial-delay" preference setting, so you can easily set it and
  forget it rather than doing "-r 30" on each open.

+ NcFTP's "file exists" prompt now includes options to apply the same
  action to all files in the batch.  This should alleviate concerns by
  users who didn't want to babysit NcFTP through a large batch of files
  which exist locally.

+ Work around Y2K bug in old wu-ftpd servers.

+ Work around broken Novell 5.00 ftp servers.

+ Fixed a getline problem where entering a character with the high-bit
  set would cause NcFTP to exit.

+ NcFTP now no longer complains when you redirect a file into it as stdin.

+ NcFTP's getline now recognizes alternate escape sequences for the arrow
  keys.

+ NcFTP now allows a user to cancel an mget at the "file exists" prompt.

+ Continue redialing if the server returns a bad user reply, since
  some servers do their "too many users logged in check" at that time.

+ Try harder to display remote connect banner.

+ Fixed bug in remote mkdir which would fail if the subdirectory was in
  the root directory.

+ A few small memory leaks plugged.

+ Try to remove extraneous slashes from paths when possible, since Win32
  doesn't like them.

+ Win32 version tries harder to sleep the actual number of seconds
  requested even if means busy waiting, since Sleep() may return
  immediately if the system has nothing else to do.

+ Win32 version now checks for a HOME or NCFTPDIR environment variable.

+ Win32 version of ncftpbookmarks no longer locks up when trying to
  send a bookmark back to NcFTP.

+ The firewall password no longer echoes to the screen on the Win32 version.

+ NcFTP tries harder to get the actual screen width so that getline
  can use the whole line for input.  ncftpbookmarks has to be built
  and installed where NcFTP can find it for this ugly hack to work.

+ NcFTP now sends your anonymous password to batch jobs, rather than
  having NcFTPBatch compute a default value.

+ Shell utilities now default to PASV-then-PORT mode like NcFTP does,
  rather than use PORT only.

+ NcFTPGet and Put are now better about reporting a server error,
  rather than simply printing the worthless "could not start data
  transfer" message.

+ Getline now handles emacs ^W sequence.


Beta21, 1999-10-03:

+ Included HTML versions of the manual pages.  This is mostly for the
  Win32 port.

+ Bug fixed in ncftpls which was printing lines with \r\n instead of
  just \n.

+ NcFTP 3 is now covered by the Artistic License.  See the LICENSE.txt
  file for details.

+ The "prefs" file is now called "prefs_v3" for better coexistence with
  NcFTP 2.


Beta20, 1999-09-29:

+ Win32 (Windows 95/98/NT) supported natively!  You can download a pre-fab
  Setup, or build it yourself from the same source base that is used for
  UNIX.

+ Readline is out, Getline is in!

+ Using my "sio" library in conjunction with LibNcFTP.  This was done
to be consistent with other platforms, which don't use alarm() and
signal handline.

+ Change of licensing.  Specifically, GPL was shown the door.  NcFTP is,
has always been, and will continue to be free software.

+ Firewall type 6 bug fixed.

+ "put -f" fixed.

+ When the server asks for a password, NcFTP will print the text, in case
  you need that for a one-time password scheme like S/Key.

+ "get -R" asks for all files by using the equivalent of /bin/ls -lRa on
  the remote server.

+ A couple of small fixes for Cygwin, although in light of native support
  for Windows this seems moot.


Beta19, 1999-06-09:

+ Changed behavior so when auto-resume is off (still the default)
  NcFTP will automatically continue without prompting if the remote
  file's size and timestamp exactly match an existing local file.
  This will make auto-resume=no a more friendly default, since you'll
  still get prompted when the files differ, but won't bother you
  like beta 18 did if the files are the same.

+ Bug fixed where "get -f" on an existing file would still have the
  existing file prompt come up.

+ Work around bug with "-s" linker flag on IRIX 6.5.

+ Small fixes to support Mac OS X Server.

+ Using inet_ntop function if available to work around IRIX 
  DNS resolution problems.

+ NcFTP's "rm -r" fixed to better handle empty directories and .dotfiles.

+ Using SITE RETRBUFSIZE/RBUFSIZ/STORBUFSIZE/SBUFSIZ commands, if
  available, to try and coordinate large TCP window support between
  client and server.

+ Special case the IIS FTP server, so if it prints directory listings
  in DOS mode, then NcFTP can parse them for file completion.

+ A few more changes for Cygwin/32.

+ Workaround definition of "lines" by AIX headers.

+ Yet another permutation of firewall added.

+ NcFTP 3 now keeps a simple transfer log like NcFTP 2 did.

+ Bug fixed in ncftpbatch which was looping if the file to get already
  existed and was identical to the remote file.

+ Local commands now work better with wildcards.

+ Bug fixed where one of the firewall handlers was not prompting you
  for your firewall password.


sh_util/ncftpls.c
/* ncftpls.c
 *
 * Copyright (c) 1999-2000 Mike Gleason, NCEMRSoft.
 * All rights reserved.
 *
 * A non-interactive utility to list directories on a remote FTP server.
 * Very useful in shell scripts!
 */

#include "syshdrs.h"

#if defined(WIN32) || defined(_WINDOWS)
#	include "..\ncftp\getopt.h"
#	define getopt Getopt
#	define optarg gOptArg
#	define optind gOptInd
	WSADATA wsaData;
	int wsaInit = 0;

	__inline void DisposeWinsock(int aUNUSED) { if (wsaInit > 0) WSACleanup(); wsaInit--; }
#	include "..\ncftp\util.h"
#	include "..\ncftp\spool.h"
#	include "..\ncftp\pref.h"
#else
#	include "../ncftp/util.h"
#	include "../ncftp/spool.h"
#	include "../ncftp/pref.h"
#endif

#include "gpshare.h"


FTPLibraryInfo gLib;
FTPConnectionInfo fi;

extern int gFirewallType;
extern char gFirewallHost[64];
extern char gFirewallUser[32];
extern char gFirewallPass[32];
extern unsigned int gFirewallPort;
extern char gFirewallExceptionList[256];
extern int gFwDataPortMode;
extern char gOS[], gVersion[];

extern char *optarg;
extern int optind;

static void
Usage(void)
{
	FILE *fp;

	fp = OpenPager();
	(void) fprintf(fp, "NcFTPLs %.5s\n\n", gVersion + 11);
	(void) fprintf(fp, "Usages:\n");
	(void) fprintf(fp, "  ncftpls [FTP flags] [-x \"ls flags\"] ftp://url.style.host/path/name/\n");
	(void) fprintf(fp, "\nls Flags:\n\
  -1     Most basic format, one item per line.\n\
  -l     Long list format.\n\
  -x XX  Other flags to pass on to the remote server.\n");
	(void) fprintf(fp, "\nFTP Flags:\n\
  -u XX  Use username XX instead of anonymous.\n\
  -p XX  Use password XX with the username.\n\
  -P XX  Use port number XX instead of the default FTP service port (21).\n\
  -j XX  Use account XX with the account (deprecated).\n\
  -d XX  Use the file XX for debug logging.\n");
	(void) fprintf(fp, "\
  -t XX  Timeout after XX seconds.\n\
  -f XX  Read the file XX for user and password information.\n\
  -E     Use regular (PORT) data connections.\n\
  -F     Use passive (PASV) data connections (default).\n\
  -r XX  Redial XX times until connected.\n");
	(void) fprintf(fp, "\nExamples:\n\
  ncftpls ftp://ftp.wustl.edu/pub/\n\
  ncftpls -1 ftp://ftp.wustl.edu/pub/\n\
  ncftpls -x \"-lrt\" ftp://ftp.wustl.edu/pub/\n");

	(void) fprintf(fp, "\nLibrary version: %s.\n", gLibNcFTPVersion + 5);
	(void) fprintf(fp, "\nThis is a freeware program by Mike Gleason (mgleason@ncftp.com).\n");
	(void) fprintf(fp, "This was built using LibNcFTP (http://www.ncftp.com/libncftp).\n");

	ClosePager(fp);
	DisposeWinsock(0);
	exit(kExitUsage);
}	/* Usage */




static void
Abort(int sigNumUNUSED)
{
	signal(SIGINT, Abort);

	/* Hopefully the I/O operation in progress
	 * will complete, and we'll abort before
	 * it starts a new block.
	 */
	fi.cancelXfer++;

	/* If the user appears to be getting impatient,
	 * restore the default signal handler so the
	 * next ^C abends the program.
	 */
	if (fi.cancelXfer >= 2)
		signal(SIGINT, SIG_DFL);
}	/* Abort */




static void
SetLsFlags(char *dst, size_t dsize, int *longMode, const char *src)
{
	char *dlim = dst + dsize - 1;
	int i, c;

	for (i=0;;) {
		c = *src++;
		if (c == '\0')
			break;
		if (c == 'l') {
			*longMode = 1;
		} else if (c == '1') {
			*longMode = 0;
		} else if (c != '-') {
			if (c == 'C') {
				*longMode = 0;
			}
			if (i == 0) {
				if (dst < dlim)
					*dst++ = '-';
			} 
			i++;
			if (dst < dlim)
				*dst++ = (char) c;
		}
	}
	*dst = '\0';
}	/* SetLsFlags */




int
main(int argc, char **argv)
{
	int result, c;
	FTPConnectionInfo savedfi;
	FTPConnectionInfo startfi;
	const char * volatile errstr;
	volatile ExitStatus es;
	char url[256];
	char urlfile[128];
	char rootcwd[256];
	int longMode = 0;
	volatile int i;
	char lsflag[32] = "";
	LineList cdlist;
	LinePtr lp;
	int rc;
	volatile int ndirs;
	char *password;

	InitWinsock();
#ifdef SIGPOLL
	NcSignal(SIGPOLL, (FTPSigProc) SIG_IGN);
#endif
	result = FTPInitLibrary(&gLib);
	if (result < 0) {
		(void) fprintf(stderr, "ncftpls: init library error %d (%s).\n", result, FTPStrError(result));
		DisposeWinsock(0);
		exit(kExitInitLibraryFailed);
	}
	result = FTPInitConnectionInfo(&gLib, &fi, kDefaultFTPBufSize);
	if (result < 0) {
		(void) fprintf(stderr, "ncftpls: init connection info error %d (%s).\n", result, FTPStrError(result));
		DisposeWinsock(0);
		exit(kExitInitConnInfoFailed);
	}

	InitUserInfo();
	fi.dataPortMode = kFallBackToSendPortMode;
	LoadFirewallPrefs(0);
	if (gFwDataPortMode >= 0)
		fi.dataPortMode = gFwDataPortMode;
	fi.debugLog = NULL;
	fi.errLog = stderr;
	fi.xferTimeout = 60 * 60;
	fi.connTimeout = 30;
	fi.ctrlTimeout = 135;
	(void) STRNCPY(fi.user, "anonymous");
	fi.host[0] = '\0';
	urlfile[0] = '\0';
	InitLineList(&cdlist);
	SetLsFlags(lsflag, sizeof(lsflag), &longMode, "-CF");
	es = kExitSuccess;

	while ((c = getopt(argc, argv, "1lx:P:u:j:p:e:d:t:r:f:EF")) > 0) switch(c) {
		case 'P':
			fi.port = atoi(optarg);	
			break;
		case 'u':
			(void) STRNCPY(fi.user, optarg);
			memset(optarg, '*', strlen(fi.user));
			break;
		case 'j':
			(void) STRNCPY(fi.acct, optarg);
			memset(optarg, '*', strlen(fi.acct));
			break;
		case 'p':
			(void) STRNCPY(fi.pass, optarg);	/* Don't recommend doing this! */
			memset(optarg, '*', strlen(fi.pass));
			break;
		case 'e':
			if (strcmp(optarg, "stdout") == 0)
				fi.errLog = stdout;
			else if (optarg[0] == '-')
				fi.errLog = stdout;
			else if (strcmp(optarg, "stderr") == 0)
				fi.errLog = stderr;
			else
				fi.errLog = fopen(optarg, "a");
			break;
		case 'd':
			if (strcmp(optarg, "stdout") == 0)
				fi.debugLog = stdout;
			else if (optarg[0] == '-')
				fi.debugLog = stdout;
			else if (strcmp(optarg, "stderr") == 0)
				fi.debugLog = stderr;
			else
				fi.debugLog = fopen(optarg, "a");
			break;
		case 't':
			SetTimeouts(&fi, optarg);
			break;
		case 'r':
			SetRedial(&fi, optarg);
			break;
		case 'f':
			ReadConfigFile(optarg, &fi);
			break;
		case 'E':
			fi.dataPortMode = kSendPortMode;
			break;
		case 'F':
			fi.dataPortMode = kPassiveMode;
			break;
		case 'l':
			SetLsFlags(lsflag, sizeof(lsflag), &longMode, "-l");
			break;
		case '1':
			SetLsFlags(lsflag, sizeof(lsflag), &longMode, "-1");
			break;
		case 'x':
			SetLsFlags(lsflag, sizeof(lsflag), &longMode, optarg);
			break;
		default:
			Usage();
	}
	if (optind > argc - 1)
		Usage();

	InitOurDirectory();

	startfi = fi;
	memset(&savedfi, 0, sizeof(savedfi));
	ndirs = argc - optind;
	for (i=optind; i<argc; i++) {
		fi = startfi;
		(void) STRNCPY(url, argv[i]);
		rc = FTPDecodeURL(&fi, url, &cdlist, urlfile, sizeof(urlfile), (int *) 0, NULL);
		(void) STRNCPY(url, argv[i]);
		if (rc == kMalformedURL) {
			(void) fprintf(stderr, "Malformed URL: %s\n", url);
			DisposeWinsock(0);
			exit(kExitMalformedURL);
		} else if (rc == kNotURL) {
			(void) fprintf(stderr, "Not a URL: %s\n", url);
			DisposeWinsock(0);
			exit(kExitMalformedURL);
		} else if (urlfile[0] != '\0') {
			/* It not obviously a directory, and they didn't say -R. */
			(void) fprintf(stderr, "Not a directory URL: %s\n", url);
			DisposeWinsock(0);
			exit(kExitMalformedURL);
		}
		
		if ((strcmp(fi.host, savedfi.host) == 0) && (strcmp(fi.user, savedfi.user) == 0)) {
			fi = savedfi;
			
			/* This host is currently open, so keep using it. */
			if (FTPChdir(&fi, rootcwd) < 0) {
				FTPPerror(&fi, fi.errNo, kErrCWDFailed, "ncftpls: Could not chdir to", rootcwd);
				es = kExitChdirFailed;
				DisposeWinsock(0);
				exit((int) es);
			}
		} else {
			if (savedfi.connected != 0) {
				errstr = "could not close remote host";
				(void) FTPCloseHost(&savedfi);
			}
			memset(&savedfi, 0, sizeof(savedfi));
			
			if (strcmp(fi.user, "anonymous") && strcmp(fi.user, "ftp")) {
				if (fi.pass[0] == '\0') {
					password = GetPass2("Password: ");		
					if (password != NULL) {
						(void) STRNCPY(fi.pass, password);
						/* Don't leave cleartext password in memory. */
						(void) memset(password, 0, strlen(fi.pass));
					}
				}
			}
			
			if (MayUseFirewall(fi.host, gFirewallType, gFirewallExceptionList) != 0) {
				fi.firewallType = gFirewallType; 
				(void) STRNCPY(fi.firewallHost, gFirewallHost);
				(void) STRNCPY(fi.firewallUser, gFirewallUser);
				(void) STRNCPY(fi.firewallPass, gFirewallPass);
				fi.firewallPort = gFirewallPort;
			}
			
			es = kExitOpenTimedOut;
			errstr = "could not open remote host";
			if ((result = FTPOpenHost(&fi)) < 0) {
				(void) fprintf(stderr, "ncftpls: cannot open %s: %s.\n", fi.host, FTPStrError(result));
				es = kExitOpenFailed;
				DisposeWinsock(0);
				exit((int) es);
			}

			if (fi.hasCLNT != kCommandNotAvailable)
				(void) FTPCmd(&fi, "CLNT NcFTPLs %.5s %s", gVersion + 11, gOS);
			
			errstr = "could not get current remote working directory from remote host";
			if (FTPGetCWD(&fi, rootcwd, sizeof(rootcwd)) < 0) {
				FTPPerror(&fi, fi.errNo, kErrPWDFailed, "ncftpls", errstr);
				es = kExitChdirFailed;
				DisposeWinsock(0);
				exit((int) es);
			}			
		}
		
		errstr = "could not change directory on remote host";
		es = kExitChdirTimedOut;
		for (lp = cdlist.first; lp != NULL; lp = lp->next) {
			if (FTPChdir(&fi, lp->line) != 0) {
				FTPPerror(&fi, fi.errNo, kErrCWDFailed, "ncftpls: Could not chdir to", lp->line);
				es = kExitChdirFailed;
				DisposeWinsock(0);
				exit((int) es);
			}
		}
		
		if (ndirs > 1) {
			fprintf(stdout, "%s%s\n\n",
				(i > optind) ? "\n\n\n" : "", url);
		}
		fflush(stdout);
		
		errstr = "could not read file from remote host";
		es = kExitXferTimedOut;
		(void) signal(SIGINT, Abort);
		if (FTPList(&fi, STDOUT_FILENO, longMode, lsflag) < 0) {
			(void) fprintf(stderr, "ncftpls: directory listing error: %s.\n", FTPStrError(fi.errNo));
			es = kExitXferFailed;
		} else {
			es = kExitSuccess;
			savedfi = fi;
		}
		(void) signal(SIGINT, SIG_DFL);
	}
	
	errstr = "could not close remote host";
	(void) FTPCloseHost(&fi);

	DisposeWinsock(0);
	exit((int) es);
}	/* main */
